ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +187,81 @@ public void shutdown (boolean async)
+     *     * The Transaction Reaper finishes monitoring all in-flight transactions
+     *     * All in-flight transactions without a timeout are beyond the prepare phase of the 2PC protocol;
+     *     to check when this condition is verified, com.arjuna.ats.arjuna.coordinator.ActionManager can be employed
+     * </p>
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     *              In case <code>isWaitForFinalRecovery() == true</code>, this parameter is override.
+     * @return the previous mode before attempting the suspension
      */
 
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForFinalRecovery checks whether suspension should delay until RecoveryModules
+            * implementing SuspendBlockingRecoveryModule recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
+
+               /*
+                * Now, it is finally possible to start checking if there are transactions that
+                * are still in need of recovery (or completion, in case of heuristics).
+                * The Action Manager gets checked as in-flight transactions with timeout zero
+                * can still be around (as the Transaction Reaper doesn't handle those txns)
+                */
+               while (_blockSuspension) {
+                   /*
+                    * There might be another thread (the embedded thread or a user-request scan)
+                    * running the recovery cycle.
+                    * It would be possible to use again doScanningWait() here to wait the end of
+                    * the ongoing scan, but the best bet is to wait for the end of the next
+                    * recycle scan with doPeriodicWait()
+                    */
+                   doPeriodicWait();
+               }
+           }
+
+           // only switch and kick everyone if we are currently ENABLED
            if (currentMode == Mode.ENABLED) {
                if (tsLogger.logger.isDebugEnabled()) {
                    tsLogger.logger.debug("PeriodicRecovery: Mode <== SUSPENDED");
                }
                setMode(Mode.SUSPENDED);
                _stateLock.notifyAll();
            }
-           if (!async) {
+
+           // At this point, in case isWaitForFinalRecovery is equal to true, an ongoing scan
```

Comment:
When you say "... an ongoing scan", is that because the run thread is still running periodic recovery scans?